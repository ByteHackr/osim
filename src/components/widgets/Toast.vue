<script setup lang="ts">

import {computed, onBeforeUnmount, ref, toRef, toRefs, watch, watchEffect} from 'vue';
import { DateTime } from 'luxon';
import ProgressRing from "@/components/widgets/ProgressRing.vue";

const props = defineProps<{
  title?: string,
  body: string,
  timestamp: DateTime,
  timeoutMs?: number,
  css?: 'primary' | 'secondary' | 'success' | 'danger' | 'warning' | 'info' | 'light' | 'dark',
}>();

const emit = defineEmits<{
  close: [],
}>();

const css = computed(() => {
  return props.css ?? 'light';
});

const active = ref<boolean>(false);
const inactiveDate = ref(Date.now());
const percentTimeRemaining = ref<number>(100);

if (props.timeoutMs) {
  const countdown: number = setInterval(() => {
    if (active.value || props.timeoutMs == null) {
      percentTimeRemaining.value = 100;
      return;
    }
    percentTimeRemaining.value = 100 - 100 * (Date.now() - inactiveDate.value) / props.timeoutMs;
  }, 33); // 1000ms / 30fps = 33ms
  onBeforeUnmount(() => {
    clearInterval(countdown);
  });
  watch(percentTimeRemaining, () => {
    if (percentTimeRemaining.value <= 0) {
      emit('close');
    }
  });
}

// Initialize ref
const timestampRelative = ref<string>(props.timestamp.toRelative() || '');

// Update function
const updateTimestamp = () => {
  timestampRelative.value = props.timestamp.toRelative() || '';
}

const updateTimestampSecond = (updatesRemaining: number = 60) => {
  updateTimestamp();
  if (updatesRemaining > 0) {
    setTimeout(updateTimestampSecond.bind(null, updatesRemaining - 1), 1000);
  } else {
    updateTimestampMinute();
  }
}

let relativeTimeInterval: number;
const updateTimestampMinute = (updatesRemaining: number = 60) => {
  updateTimestamp();
  if (updatesRemaining > 0) {
    setTimeout(updateTimestampMinute.bind(null, updatesRemaining - 1), 60000);
  } else {
    relativeTimeInterval = setInterval(updateTimestamp, 60000 * 5); // 5 minutes
  }
}
updateTimestampSecond();

onBeforeUnmount(() => {
  clearInterval(relativeTimeInterval);
});

watchEffect(() => {
  updateTimestamp();
});

const transitionDurationMs = ref(16);

</script>

<template>
  <div
      class="osim-toast toast show"
      :class="['text-bg-' + css]"
      role="alert"
      aria-labelledby="modalTitle"
      aria-live="assertive"
      aria-atomic="true"
      @mouseover="active = true; transitionDurationMs = 0"
      @mouseleave="inactiveDate = Date.now(); transitionDurationMs = 16; active = false"
  >
    <!--:style="{display: show? 'block' : 'none'}"-->
    <div class="toast-header">
      <slot name="header">
        <ProgressRing
            class="osim-toast-timer"
            v-if="timeoutMs"
            color="grey"
            :progress="percentTimeRemaining"
            :diameter="12"
            :stroke="6"
            :transition-duration-ms="transitionDurationMs"
            direction="down"
        />
        <strong class="me-auto">{{ title ?? '' }}</strong>
        <small class="text-muted">{{ timestampRelative }}</small>
        <button type="button" class="btn-close" aria-label="Close" @click="$emit('close')"></button>
      </slot>
    </div>
    <div class="toast-body osim-toast-body">
      <slot name="body">
        {{ body }}
      </slot>
    </div>
  </div>
</template>

<style scoped>
.osim-toast {
  min-width: var(--bs-toast-max-width);
}
.osim-toast-body {
  white-space: pre-wrap;
}
.osim-toast-timer {
  margin-right: 5px;
}
/*
.osim-body-html {
  all: initial;
}
*/
</style>
