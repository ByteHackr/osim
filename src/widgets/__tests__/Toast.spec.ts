import { mount, flushPromises } from '@vue/test-utils';
import { createTestingPinia } from '@pinia/testing';
import { DateTime } from 'luxon';

import Toast, { type ToastProps } from '@/widgets/Toast/Toast.vue';

describe('toast', () => {
  const mountToast = (props?: Partial<ToastProps>) => {
    return mount(Toast, {
      props: {
        timestamp: DateTime.fromISO('2024-08-28T10:10:10.000Z'),
        body: 'Test body',
        ...props,
      },
      global: {
        plugins: [createTestingPinia()],
      },
    });
  };

  beforeEach(() => {
    vi.useFakeTimers({
      now: new Date('2024-08-28T11:10:10.000Z').getTime(),
    });
  });

  afterEach(() => {
    vi.useRealTimers();
  });

  it.each<{ body: string; css: ToastProps['css']; title: string }>([
    { title: 'Success', body: 'Test body', css: 'success' },
    { title: 'Error', body: 'Test body', css: 'danger' },
    { title: 'Warn', body: 'Test body', css: 'warning' },
    { title: 'Default', body: 'Test body', css: undefined },
  ])('should render component with title $title, body $body and $css style', ({ body, css, title }) => {
    const wrapper = mountToast({ title, body, css });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should close toast when close button is clicked', async () => {
    const wrapper = mountToast();

    await wrapper.find('.btn-close').trigger('click');

    expect(wrapper.emitted('close')).toBeTruthy();
  });

  it('should close toast after timeoutMs', async () => {
    const wrapper = mountToast({ timeoutMs: 500 });

    vi.advanceTimersByTime(1000);
    await flushPromises();

    expect(wrapper.emitted('close')).toBeTruthy();
    expect(wrapper.emitted('stale')).toBeFalsy();
  });

  it('should emit stale event after freshMs', async () => {
    const wrapper = mountToast({ timeoutMs: 50000 });

    vi.advanceTimersByTime(11000);
    await flushPromises();

    expect(wrapper.emitted('close')).toBeFalsy();
    expect(wrapper.emitted('stale')).toBeTruthy();
  });
});
