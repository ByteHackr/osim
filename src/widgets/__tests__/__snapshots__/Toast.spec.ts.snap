// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`toast > should render component with title 'Default', body 'Test body' and undefined style 1`] = `
"<div data-v-2b92ce91="" class="osim-toast toast show text-bg-light fresh-leave-active" role="alert" aria-labelledby="modalTitle" aria-live="assertive" aria-atomic="true">
  <!--:style="{display: show? 'block' : 'none'}"-->
  <div data-v-2b92ce91="" class="toast-header"><strong data-v-2b92ce91="" class="me-auto">Default</strong><small data-v-2b92ce91="" class="text-muted">1 hour ago</small><button data-v-2b92ce91="" type="button" class="osim-toast-close-btn btn-close" aria-label="Close">
      <!--v-if-->
    </button></div>
  <div data-v-2b92ce91="" class="toast-body osim-toast-body">
    <!-- eslint-disable-next-line vue/no-v-html -->
    <!--v-if-->
    <div data-v-2b92ce91="">Test body</div>
  </div>
</div>"
`;

exports[`toast > should render component with title 'Error', body 'Test body' and 'danger' style 1`] = `
"<div data-v-2b92ce91="" class="osim-toast toast show text-bg-danger fresh-leave-active" role="alert" aria-labelledby="modalTitle" aria-live="assertive" aria-atomic="true">
  <!--:style="{display: show? 'block' : 'none'}"-->
  <div data-v-2b92ce91="" class="toast-header"><strong data-v-2b92ce91="" class="me-auto">Error</strong><small data-v-2b92ce91="" class="text-muted">1 hour ago</small><button data-v-2b92ce91="" type="button" class="osim-toast-close-btn btn-close" aria-label="Close">
      <!--v-if-->
    </button></div>
  <div data-v-2b92ce91="" class="toast-body osim-toast-body">
    <!-- eslint-disable-next-line vue/no-v-html -->
    <!--v-if-->
    <div data-v-2b92ce91="">Test body</div>
  </div>
</div>"
`;

exports[`toast > should render component with title 'Success', body 'Test body' and 'success' style 1`] = `
"<div data-v-2b92ce91="" class="osim-toast toast show text-bg-success fresh-leave-active" role="alert" aria-labelledby="modalTitle" aria-live="assertive" aria-atomic="true">
  <!--:style="{display: show? 'block' : 'none'}"-->
  <div data-v-2b92ce91="" class="toast-header"><strong data-v-2b92ce91="" class="me-auto">Success</strong><small data-v-2b92ce91="" class="text-muted">1 hour ago</small><button data-v-2b92ce91="" type="button" class="osim-toast-close-btn btn-close" aria-label="Close">
      <!--v-if-->
    </button></div>
  <div data-v-2b92ce91="" class="toast-body osim-toast-body">
    <!-- eslint-disable-next-line vue/no-v-html -->
    <!--v-if-->
    <div data-v-2b92ce91="">Test body</div>
  </div>
</div>"
`;

exports[`toast > should render component with title 'Warn', body 'Test body' and 'warning' style 1`] = `
"<div data-v-2b92ce91="" class="osim-toast toast show text-bg-warning fresh-leave-active" role="alert" aria-labelledby="modalTitle" aria-live="assertive" aria-atomic="true">
  <!--:style="{display: show? 'block' : 'none'}"-->
  <div data-v-2b92ce91="" class="toast-header"><strong data-v-2b92ce91="" class="me-auto">Warn</strong><small data-v-2b92ce91="" class="text-muted">1 hour ago</small><button data-v-2b92ce91="" type="button" class="osim-toast-close-btn btn-close" aria-label="Close">
      <!--v-if-->
    </button></div>
  <div data-v-2b92ce91="" class="toast-body osim-toast-body">
    <!-- eslint-disable-next-line vue/no-v-html -->
    <!--v-if-->
    <div data-v-2b92ce91="">Test body</div>
  </div>
</div>"
`;
